### PR TITLE
pcie, new case for invalid PCIe slots

### PR DESCRIPTION
**Pull request by vasekhodina**
This case adds checks for devices attached to PCIe controller that have
slot number bigger than one or something else than a number.
    
VIRT-48333

### CODE REVIEW

**Comment by vasekhodina** [comment]
Test results from RHEL 9.2:
image

**Comment by vasekhodina** [comment]
@dzhengfy Please review this PR. Thank you.

**Comment by dzhengfy** [comment]
Please hold on this pr before the https://github.com/autotest/tp-libvirt/pull/4573 is merged because there is a critical update required. See https://github.com/autotest/tp-libvirt/pull/4573#discussion_r1016406856. @vasekhodina if you do not mind, could we change it to draft for now?

**Comment by vasekhodina** [comment]
Test results from RHEL 9.2:
```
]# avocado run --vt-type libvirt pcie_root_port_controller
No python imaging library installed. Screendump and Windows guest BSOD detection are disabled. In order to enable it, please install python-imaging or the equivalent for your distro.
No python imaging library installed. PPM image conversion to JPEG disabled. In order to enable it, please install python-imaging or the equivalent for your distro.
JOB ID     : b9c62cb1f8fb3897166f710c893496d4d95d3ef0
JOB LOG    : /var/lib/avocado/job-results/job-2022-12-20T09.57-b9c62cb/job.log
 (1/7) type_specific.io-github-autotest-libvirt.pcie_root_port_controller.positive_test.controllers_different_chassis_same_port: STARTED
 (1/7) type_specific.io-github-autotest-libvirt.pcie_root_port_controller.positive_test.controllers_different_chassis_same_port: PASS (54.48 s)
 (2/7) type_specific.io-github-autotest-libvirt.pcie_root_port_controller.negative_test.device_slot_invalid: STARTED
 (2/7) type_specific.io-github-autotest-libvirt.pcie_root_port_controller.negative_test.device_slot_invalid: PASS (6.09 s)
 (3/7) type_specific.io-github-autotest-libvirt.pcie_root_port_controller.negative_test.device_slot_wrong_number: STARTED
 (3/7) type_specific.io-github-autotest-libvirt.pcie_root_port_controller.negative_test.device_slot_wrong_number: PASS (6.12 s)
 (4/7) type_specific.io-github-autotest-libvirt.pcie_root_port_controller.negative_test.controllers_same_chassis_same_port: STARTED
 (4/7) type_specific.io-github-autotest-libvirt.pcie_root_port_controller.negative_test.controllers_same_chassis_same_port: PASS (7.40 s)
 (5/7) type_specific.io-github-autotest-libvirt.pcie_root_port_controller.negative_test.controllers_same_chassis_different_port: STARTED
 (5/7) type_specific.io-github-autotest-libvirt.pcie_root_port_controller.negative_test.controllers_same_chassis_different_port: PASS (8.09 s)
 (6/7) type_specific.io-github-autotest-libvirt.pcie_root_port_controller.negative_test.index_equals_address_bus: STARTED
 (6/7) type_specific.io-github-autotest-libvirt.pcie_root_port_controller.negative_test.index_equals_address_bus: PASS (5.92 s)
 (7/7) type_specific.io-github-autotest-libvirt.pcie_root_port_controller.negative_test.index_less_than_address_bus: STARTED
 (7/7) type_specific.io-github-autotest-libvirt.pcie_root_port_controller.negative_test.index_less_than_address_bus: PASS (5.95 s)
RESULTS    : PASS 7 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /var/lib/avocado/job-results/job-2022-12-20T09.57-b9c62cb/results.html
JOB TIME   : 102.15 s
```

File: libvirt/tests/cfg/controller/pcie_root_port_controller.cfg [hunk 1]
```diff
@@ -13,9 +13,29 @@
...
+                    interface_slot = 1
+                    minimal_interface_dict = '{"mac_address": "52:54:00:15:04:28", "source": {"network": "default"}}'
+                    failure_message = "Invalid PCI address .* slot must be <= 0"
                 - controllers_same_chassis_same_port:
                     second_controller_model = "pcie-root-port"
                     second_controller_target = "{'chassis':1,'port':'0x8'}"
                 - controllers_same_chassis_different_port:
                     second_controller_model = "pcie-root-port"
                     second_controller_target = "{'chassis':1,'port':'0x4'}"
+                - index_equals_address_bus:
```
Comment (dzhengfy) [inline libvirt/tests/cfg/controller/pcie_root_port_controller.cfg:33]: L33 to Line 41 should not belong to this PR. Please elaborate it.

Comment (vasekhodina) [reply libvirt/tests/cfg/controller/pcie_root_port_controller.cfg]: I've removed the lines. I forgot to delete them when I was merging other commits from branches that are modifying the same file.

**Comment by dzhengfy** [comment]
@vasekhodina please rebase.

File: libvirt/tests/src/controller/pcie_root_port_controller.py [hunk 2]
```diff
@@ -97,29 +105,60 @@ def check_define_vm(vmxml, params, test):
...
+    :param minimal_dict: Dictionary with interface values from config file
+    :interface bus: Optional value for address bus
+    :interface_slot: Optional value for address slot to update
+    """
+    interface_address = {"attrs": {}}
+    if interface_bus:
+        interface_address["attrs"].update({"bus": interface_bus})
+    if interface_slot:
+        interface_address["attrs"].update({"slot": interface_slot})
+    #interface_address = {"attrs": {"bus": interface_bus, "slot": interface_slot}}
```
Comment (dzhengfy) [inline libvirt/tests/src/controller/pcie_root_port_controller.py:156]: Please remove the commented line.

File: libvirt/tests/src/controller/pcie_root_port_controller.py [hunk 3]
```diff
@@ -97,29 +105,60 @@ def check_define_vm(vmxml, params, test):
...
     if failure_message:
         libvirt.check_result(cmd_result, [failure_message])
     else:
         libvirt.check_exit_status(cmd_result, status_error)
 
 
+def customize_interface_dict(minimal_dict, interface_bus=None, interface_slot=None):
+    """
+    Update interface dictionary and prepare it for adding to VM XML.
+    :param minimal_dict: Dictionary with interface values from config file
```
Comment (dzhengfy) [inline libvirt/tests/src/controller/pcie_root_port_controller.py:147]: Add one blank line above this line.

**Review by dzhengfy — COMMENTED** [review]
Others LGTM

File: libvirt/tests/src/controller/pcie_root_port_controller.py [hunk 4]
```diff
@@ -22,13 +26,17 @@ def setup_test(params):
     second_controller_target = ast.literal_eval(
         params.get("second_controller_target", "{}"))
     vmxml = vm_xml.VMXML.new_from_inactive_dumpxml(params.get("main_vm", "avocado-vt-vm1"))
+    test_define_only = params.get("test_define_only")
```
Comment (Yingshun) [inline libvirt/tests/src/controller/pcie_root_port_controller.py:29]: Is it better the check if 'test_define_only' in cfg is 'yes'?

File: libvirt/tests/cfg/controller/pcie_root_port_controller.cfg [hunk 5]
```diff
@@ -13,6 +13,17 @@
         - negative_test:
             status_error = "yes"
             variants:
+                - device_slot_invalid:
+                    test_define_only = "yes"
+                    interface_slot = "a"
+                    minimal_interface_dict = '{"mac_address": "52:54:00:15:04:28", "source": {"network": "default"}}'
```
Comment (Yingshun) [inline libvirt/tests/cfg/controller/pcie_root_port_controller.cfg:19]: I wonder if the 'mac_address' should be randomly generated? Can we avoid hard-coding here?

File: libvirt/tests/src/controller/pcie_root_port_controller.py [hunk 6]
```diff
@@ -1,8 +1,14 @@
-import logging
+"""
+Module automating these manual test cases:
+VIRT-48333,VIRT-54001,VIRT-55416
```
[the remark this answers is not in the document]
Comment (Yingshun) [reply libvirt/tests/src/controller/pcie_root_port_controller.py:3]: +1

File: libvirt/tests/src/controller/pcie_root_port_controller.py [hunk 7]
```diff
@@ -97,29 +105,60 @@ def check_define_vm(vmxml, params, test):
...
+    :interface bus: Optional value for address bus
+    :interface_slot: Optional value for address slot to update
+    """
+    interface_address = {"attrs": {}}
+    if interface_bus:
+        interface_address["attrs"].update({"bus": interface_bus})
+    if interface_slot:
+        interface_address["attrs"].update({"slot": interface_slot})
+    #interface_address = {"attrs": {"bus": interface_bus, "slot": interface_slot}}
+    minimal_dict.update({"address": interface_address})
```
Comment (Yingshun) [inline libvirt/tests/src/controller/pcie_root_port_controller.py:157]: What if 'interface_address' is {"attrs": {}}?

File: libvirt/tests/cfg/controller/pcie_root_port_controller.cfg (hunk 5 above)
Comment (dzhengfy) [reply libvirt/tests/cfg/controller/pcie_root_port_controller.cfg:19]: @Yingshun Do we have existing functions to randomly generate the mac address?

Comment (Yingshun) [reply libvirt/tests/cfg/controller/pcie_root_port_controller.cfg:19]: How about using utils_net.generate_mac_address_simple()?

File: libvirt/tests/src/controller/pcie_root_port_controller.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (vasekhodina) [reply libvirt/tests/src/controller/pcie_root_port_controller.py:3]: Removed. I hoped it would help finding which automated cases belong to manual cases.

File: libvirt/tests/src/controller/pcie_root_port_controller.py (hunk 2 above)
Comment (vasekhodina) [reply libvirt/tests/src/controller/pcie_root_port_controller.py:156]: Removed.

File: libvirt/tests/src/controller/pcie_root_port_controller.py (hunk 3 above)
Comment (vasekhodina) [reply libvirt/tests/src/controller/pcie_root_port_controller.py:147]: Line added.

File: libvirt/tests/src/controller/pcie_root_port_controller.py (hunk 4 above)
Comment (vasekhodina) [reply libvirt/tests/src/controller/pcie_root_port_controller.py:29]: Fixed.

File: libvirt/tests/src/controller/pcie_root_port_controller.py (hunk 7 above)
Comment (vasekhodina) [reply libvirt/tests/src/controller/pcie_root_port_controller.py:157]: An empty address element will be created: '<address type="pci" \>'. I have added an if statement so that the interface_address is created only when the 2 last arguments are not null.

File: libvirt/tests/cfg/controller/pcie_root_port_controller.cfg (hunk 5 above)
Comment (Yingshun) [reply libvirt/tests/cfg/controller/pcie_root_port_controller.cfg:19]: @vasekhodina  any updates?

**Review by Yingshun — COMMENTED** [review]
Others LGTM

File: libvirt/tests/cfg/controller/pcie_root_port_controller.cfg [hunk 8]
```diff
@@ -13,6 +13,17 @@
...
+                    failure_message = "Invalid value for attribute 'slot' in element 'address': 'a'. Expected .*integer value"
+                - device_slot_wrong_number:
+                    test_define_only = "yes"
+                    interface_slot_type = "hex"
+                    interface_slot = 1
+                    minimal_interface_dict = '{"mac_address": "52:54:00:15:04:28", "source": {"network": "default"}}'
```
Comment (dzhengfy) [inline libvirt/tests/cfg/controller/pcie_root_port_controller.cfg:25]: Same as above

File: libvirt/tests/cfg/controller/pcie_root_port_controller.cfg (hunk 5 above)
Comment (vasekhodina) [reply libvirt/tests/cfg/controller/pcie_root_port_controller.cfg:19]: Oh, sorry I have thought about this. The case doesn't mention that the MAC address needs to be randomly generated, so I assumed only one is ok. Do you think it needs to be generated every run? @Yingshun @dzhengfy

Comment (Yingshun) [reply libvirt/tests/cfg/controller/pcie_root_port_controller.cfg:19]: I've checked virtual network cases, most of them use different mac address and only a few tests which related to 'mac address' have a fixed value in cfg, so I still think we should avoid hard-coding here. BTW, as this variant name is named 'minimal_interface_dict', can we only keep 'source' if you don't want to generate mac in the code?

Comment (vasekhodina) [reply libvirt/tests/cfg/controller/pcie_root_port_controller.cfg:19]: @Yingshun Yes, let me check. Maybe libvirt itself will generate the mac address.

File: libvirt/tests/src/controller/pcie_root_port_controller.py [hunk 9]
```diff
@@ -97,29 +101,61 @@ def check_define_vm(vmxml, params, test):
...
     else:
         libvirt.check_exit_status(cmd_result, status_error)
 
 
+def customize_interface_dict(minimal_dict, interface_bus=None, interface_slot=None):
+    """
+    Update interface dictionary and prepare it for adding to VM XML.
+
+    :param minimal_dict: Dictionary with interface values from config file
+    :interface bus: Optional value for address bus
```
Comment (dzhengfy) [inline libvirt/tests/src/controller/pcie_root_port_controller.py:145]: :param missing

File: libvirt/tests/src/controller/pcie_root_port_controller.py [hunk 10]
```diff
@@ -97,29 +101,61 @@ def check_define_vm(vmxml, params, test):
...
+    :interface_slot: Optional value for address slot to update
```
Comment (dzhengfy) [inline libvirt/tests/src/controller/pcie_root_port_controller.py:146]: :param missing

**Review by dzhengfy — COMMENTED** [review]
Others LGTM

**Review by Yingshun — APPROVED** [review]

File: libvirt/tests/src/controller/pcie_root_port_controller.py (hunk 9 above)
Comment (vasekhodina) [reply libvirt/tests/src/controller/pcie_root_port_controller.py:145]: Fixed.

File: libvirt/tests/src/controller/pcie_root_port_controller.py (hunk 10 above)
Comment (vasekhodina) [reply libvirt/tests/src/controller/pcie_root_port_controller.py:146]: Fixed.

**Comment by vasekhodina** [comment]
Tests still pass even after latest changes:
```
# avocado run --vt-type libvirt pcie_root_port_controller
No python imaging library installed. Screendump and Windows guest BSOD detection are disabled. In order to enable it, please install python-imaging or the equivalent for your distro.
No python imaging library installed. PPM image conversion to JPEG disabled. In order to enable it, please install python-imaging or the equivalent for your distro.
JOB ID     : 286a4f1ada49d52446133e3b14f8a0ccb94e974c
JOB LOG    : /var/lib/avocado/job-results/job-2023-01-31T07.10-286a4f1/job.log
 (1/7) type_specific.io-github-autotest-libvirt.pcie_root_port_controller.positive_test.controllers_different_chassis_same_port: STARTED
 (1/7) type_specific.io-github-autotest-libvirt.pcie_root_port_controller.positive_test.controllers_different_chassis_same_port: PASS (84.06 s)
 (2/7) type_specific.io-github-autotest-libvirt.pcie_root_port_controller.negative_test.device_slot_invalid: STARTED
 (2/7) type_specific.io-github-autotest-libvirt.pcie_root_port_controller.negative_test.device_slot_invalid: PASS (27.85 s)
 (3/7) type_specific.io-github-autotest-libvirt.pcie_root_port_controller.negative_test.device_slot_wrong_number: STARTED
 (3/7) type_specific.io-github-autotest-libvirt.pcie_root_port_controller.negative_test.device_slot_wrong_number: PASS (15.91 s)
 (4/7) type_specific.io-github-autotest-libvirt.pcie_root_port_controller.negative_test.controllers_same_chassis_same_port: STARTED
 (4/7) type_specific.io-github-autotest-libvirt.pcie_root_port_controller.negative_test.controllers_same_chassis_same_port: PASS (17.49 s)
 (5/7) type_specific.io-github-autotest-libvirt.pcie_root_port_controller.negative_test.controllers_same_chassis_different_port: STARTED
 (5/7) type_specific.io-github-autotest-libvirt.pcie_root_port_controller.negative_test.controllers_same_chassis_different_port: PASS (28.35 s)
 (6/7) type_specific.io-github-autotest-libvirt.pcie_root_port_controller.negative_test.index_equals_address_bus: STARTED
 (6/7) type_specific.io-github-autotest-libvirt.pcie_root_port_controller.negative_test.index_equals_address_bus: PASS (14.56 s)
 (7/7) type_specific.io-github-autotest-libvirt.pcie_root_port_controller.negative_test.index_less_than_address_bus: STARTED
 (7/7) type_specific.io-github-autotest-libvirt.pcie_root_port_controller.negative_test.index_less_than_address_bus: PASS (14.44 s)
RESULTS    : PASS 7 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /var/lib/avocado/job-results/job-2023-01-31T07.10-286a4f1/results.html
```

**Review by dzhengfy — APPROVED** [review]
